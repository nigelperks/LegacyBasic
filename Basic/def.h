// Legacy BASIC
// Copyright (c) 2022-24 Nigel Perks
// Utility functions for DEF.

#pragma once

#include "source.h"
#include "bcode.h"

struct def {
  BCODE* bcode;
  SOURCE* source; // reference to stored program source if applicable
  unsigned source_line;
};

// takes ownership of the BCODE
struct def * new_def(BCODE*, SOURCE*, unsigned source_line);

void delete_def(struct def *);
