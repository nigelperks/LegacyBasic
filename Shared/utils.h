// Legacy BASIC
// Utility types and functions.
// Copyright (c) 2021-24 Nigel Perks

#ifndef UTILS_H
#define UTILS_H

#include <stdio.h>
#include <stdbool.h>

extern const char* progname;

typedef unsigned short U16;

void fatal(const char* fmt, ...);
void error(const char* fmt, ...);

void* emalloc(size_t);
void* erealloc(void*, size_t);
void* ecalloc(size_t count, size_t size);
char* estrdup(const char*);

void efree(void*);

extern unsigned long malloc_count, free_count;

enum { TYPE_ERR, TYPE_NUM, TYPE_STR };

bool string_name(const char*);

void space(unsigned count, FILE*);

#endif // UTILS_H
