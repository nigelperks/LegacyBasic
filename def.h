#pragma once

struct def {
  unsigned params;
  unsigned source_line;
  unsigned pc;
};

struct def * new_def(unsigned params, unsigned source_line, unsigned pc);
void delete_def(struct def *);
