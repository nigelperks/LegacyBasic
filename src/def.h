#pragma once

#include "source.h"
#include "bcode.h"

struct def {
  unsigned params;
  unsigned source_line;  // 0 if immediate statement
  BCODE* code;
};

struct def * new_def(void);
void delete_def(struct def *);
