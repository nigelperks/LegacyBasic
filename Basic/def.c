// Legacy BASIC
// Copyright (c) 2022-24 Nigel Perks
// Utility functions for DEF.

#include "def.h"
#include "utils.h"

struct def * new_def(BCODE* bc, SOURCE* source, unsigned source_line) {
  struct def * def = emalloc(sizeof *def);
  def->bcode = bc;
  def->source = source;
  def->source_line = source_line;
  return def;
}

void delete_def(struct def * def) {
  if (def) {
    delete_bcode(def->bcode);
    efree(def);
  }
}
