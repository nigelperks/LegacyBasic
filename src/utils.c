// Legacy BASIC
// Utility types and functions.
// Copyright (c) 2021-2 Nigel Perks

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include <assert.h>
#include "utils.h"

const char* progname;

void fatal(const char* fmt, ...) {
  fflush(stdout);
  va_list ap;
  if (progname)
    fprintf(stderr, "%s: ", progname);
  fprintf(stderr, "fatal: ");
  va_start(ap, fmt);
  vfprintf(stderr, fmt, ap);
  va_end(ap);
  exit(EXIT_FAILURE);
}

void error(const char* fmt, ...) {
  fflush(stdout);
  va_list ap;
  va_start(ap, fmt);
  vprintf(fmt, ap);
  va_end(ap);
  putchar('\n');
}

unsigned long malloc_count;
unsigned long free_count;

void* emalloc(size_t sz) {
  void* p = malloc(sz);
  if (p == NULL)
    fatal("out of memory (emalloc)\n");
  malloc_count++;
  return p;
}

void* erealloc(void* p, size_t sz) {
  if (p)
    free_count++;
  p = realloc(p, sz);
  if (p == NULL)
    fatal("out of memory (erealloc)\n");
  malloc_count++;
  return p;
}

void* ecalloc(size_t count, size_t size) {
  void* p = calloc(count, size);
  if (p == NULL)
    fatal("out of memory (ecalloc)\n");
  malloc_count++;
  return p;
}

char* estrdup(const char* s) {
  char* t = NULL;
  if (s) {
    if (s[0]) {
      t = emalloc(strlen(s) + 1);
      strcpy(t, s);
    }
    else {
      t = emalloc(1);
      t[0] = '\0';
    }
  }
  return t;
}

void efree(void* p) {
  if (p) {
    free_count++;
    free(p);
  }
}

bool string_name(const char* name) {
  return name != NULL && name[0] && name[strlen(name) - 1] == '$';
}

void space(unsigned n, FILE* fp) {
  while (n--)
    putc(' ', fp);
}

#ifdef UNIT_TEST

#include "CuTest.h"

static void test_string_name(CuTest* tc) {
  CuAssertIntEquals(tc, false, string_name(NULL));
  CuAssertIntEquals(tc, false, string_name(""));
  CuAssertIntEquals(tc, true, string_name("$"));
  CuAssertIntEquals(tc, true, string_name("a$"));
  CuAssertIntEquals(tc, false, string_name("a"));
}

CuSuite* utils_test_suite(void) {
  CuSuite* suite = CuSuiteNew();
  SUITE_ADD_TEST(suite, test_string_name);
  return suite;
}

#endif
