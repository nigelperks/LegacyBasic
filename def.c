// Legacy BASIC
// Copyright (c) 2022 Nigel Perks
// Utility functions for DEF.

#include "def.h"
#include "utils.h"

struct def * new_def(unsigned params, unsigned source_line, unsigned pc) {
  struct def * p = emalloc(sizeof *p);
  p->params = params;
  p->source_line = source_line;
  p->pc = pc;
  return p;
}

void delete_def(struct def * def) {
  efree(def);
}

