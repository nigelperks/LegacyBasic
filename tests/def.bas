100 DEF FNQ(X)=4*X
110 PRINT FNQ(3.14),FNQ(-9)
120 DEF FNQ(X)=X*X
130 PRINT FNQ(-3)
140 DEF old$(n)="He is " + str$(n) + " years old"
150 PRINT old$(123)
