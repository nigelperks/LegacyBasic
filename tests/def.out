 12.56   -36 
 9 
He is 123 years old
